File: ios/Sources/VaultSync/Data/Sync/CrashRecoveryHandler.swift
import Foundation

public final class CrashRecoveryHandler: Sendable {
    private let syncRepository: any SyncRepository
    private let documentRepository: any DocumentRepository
    private let fileStorage: any FileStorage
    private let cryptoService: any CryptoService

    public init(
        syncRepository: any SyncRepository,
        documentRepository: any DocumentRepository,
        fileStorage: any FileStorage,
        cryptoService: any CryptoService
    ) {
        self.syncRepository = syncRepository
        self.documentRepository = documentRepository
        self.fileStorage = fileStorage
        self.cryptoService = cryptoService
    }

    public func recover() async throws -> Int {
        let runningOps = try await syncRepository.getRunningOperations()
        var recovered = 0

        for op in runningOps {
            if let doc = try await documentRepository.getDocument(id: op.documentId),
               await fileStorage.fileExists(at: doc.localPath) {
                let bytes = try await fileStorage.readFile(at: doc.localPath)
                if !bytes.isEmpty {
                    try await syncRepository.updateOperationStatus(
                        id: op.id,
                        status: .pending,
                        errorMessage: "Recovered from unexpected shutdown",
                        completedAt: nil
                    )
                    try await documentRepository.updateSyncStatus(id: doc.id, status: .pending)
                    recovered += 1
                } else {
                    try await syncRepository.updateOperationStatus(
                        id: op.id,
                        status: .failed,
                        errorMessage: "Zero-length file after crash",
                        completedAt: Date()
                    )
                    try await documentRepository.updateSyncStatus(id: doc.id, status: .failed)
                    recovered += 1
                }
            } else {
                try await syncRepository.updateOperationStatus(
                    id: op.id,
                    status: .failed,
                    errorMessage: "Target file missing after crash",
                    completedAt: Date()
                )
                recovered += 1
            }
        }
        return recovered
    }
}

File: android/app/src/main/java/com/vaultsync/data/sync/CrashRecoveryHandler.kt
package com.vaultsync.data.sync

import com.vaultsync.domain.model.OperationStatus
import com.vaultsync.domain.model.SyncStatus
import com.vaultsync.domain.repository.CryptoService
import com.vaultsync.domain.repository.DocumentRepository
import com.vaultsync.domain.repository.FileStorage
import com.vaultsync.domain.repository.SyncRepository

class CrashRecoveryHandler(
    private val syncRepository: SyncRepository,
    private val documentRepository: DocumentRepository,
    private val fileStorage: FileStorage,
    private val cryptoService: CryptoService
) {
    suspend fun recover(): Int {
        val orphanedOps = syncRepository.getRunningOperations()
        var recoveredCount = 0

        for (op in orphanedOps) {
            val doc = documentRepository.getDocument(op.documentId)
            if (doc != null && fileStorage.fileExists(doc.localPath)) {
                try {
                    val fileBytes = fileStorage.readFile(doc.localPath)
                    if (fileBytes.isNotEmpty()) {
                        // Mark document and op as PENDING to retry clean execution
                        syncRepository.updateOperationStatus(
                            id = op.id,
                            status = OperationStatus.PENDING,
                            errorMessage = "Recovered from unexpected shutdown"
                        )
                        documentRepository.updateSyncStatus(doc.id, SyncStatus.PENDING)
                        recoveredCount++
                    } else {
                        syncRepository.updateOperationStatus(
                            id = op.id,
                            status = OperationStatus.FAILED,
                            errorMessage = "Zero-length file after crash",
                            completedAt = System.currentTimeMillis()
                        )
                        documentRepository.updateSyncStatus(doc.id, SyncStatus.FAILED)
                        recoveredCount++
                    }
                } catch (e: Exception) {
                    syncRepository.updateOperationStatus(
                        id = op.id,
                        status = OperationStatus.FAILED,
                        errorMessage = "Recovery failed: ${e.message}",
                        completedAt = System.currentTimeMillis()
                    )
                    recoveredCount++
                }
            } else {
                // Document or file disappeared during crash
                syncRepository.updateOperationStatus(
                    id = op.id,
                    status = OperationStatus.FAILED,
                    errorMessage = "Target file missing after crash",
                    completedAt = System.currentTimeMillis()
                )
                recoveredCount++
            }
        }
        return recoveredCount
    }
}
